tensorflow/lite/micro/kernels/detection_postprocess.cc

```diff
@@ -0,0 +1,855 @@
+                                   int x, int y=0, int z=0) {
+  int size = 1;
+
+  size = size * x;
+  size = (y > 0) ? size * y : size;
+  size = (z > 0) ? size * z : size;
+
+  *dims = reinterpret_cast<TfLiteIntArray*>(context->AllocatePersistentBuffer(
+      context, TfLiteIntArrayGetSizeInBytes(size)));
+
+  (*dims)->size = size;
+  (*dims)->data[0] = x;
+  if (y > 0) {
+    (*dims)->data[1] = y;
+  }
+  if (z > 0) {
+    (*dims)->data[2] = z;
+  }
+
+  return kTfLiteOk;
+}
+
+void* Init(TfLiteContext* context, const char* buffer, size_t length) {
+  OpData* op_data = nullptr;
+
+  const uint8_t* buffer_t = reinterpret_cast<const uint8_t*>(buffer);
+  const flexbuffers::Map& m = flexbuffers::GetRoot(buffer_t, length).AsMap();
+
+  TFLITE_DCHECK(context->AllocatePersistentBuffer != nullptr);
+  op_data = reinterpret_cast<OpData*>(context->AllocatePersistentBuffer(context, sizeof(OpData)));
+
+  op_data->max_detections = m["max_detections"].AsInt32();
+  op_data->max_classes_per_detection = m["max_classes_per_detection"].AsInt32();
+  if (m["detections_per_class"].IsNull())
+    op_data->detections_per_class = kNumDetectionsPerClass;
+  else
+    op_data->detections_per_class = m["detections_per_class"].AsInt32();
+  if (m["use_regular_nms"].IsNull())
+    op_data->use_regular_non_max_suppression = false;
+  else
+    op_data->use_regular_non_max_suppression = m["use_regular_nms"].AsBool();
+
+  op_data->non_max_suppression_score_threshold =
+      m["nms_score_threshold"].AsFloat();
+  op_data->intersection_over_union_threshold = m["nms_iou_threshold"].AsFloat();
+  op_data->num_classes = m["num_classes"].AsInt32();
+  op_data->scale_values.y = m["y_scale"].AsFloat();
+  op_data->scale_values.x = m["x_scale"].AsFloat();
+  op_data->scale_values.h = m["h_scale"].AsFloat();
+  op_data->scale_values.w = m["w_scale"].AsFloat();
+
+  return op_data;
+}
+
+void Free(TfLiteContext* context, void* buffer) {}
+
+TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
+  auto* op_data = static_cast<OpData*>(node->user_data);
+
+  // Inputs: box_encodings, scores, anchors
+  TF_LITE_ENSURE_EQ(context, NumInputs(node), 3);
+  const TfLiteTensor* input_box_encodings =
+      GetInput(context, node, kInputTensorBoxEncodings);
+  const TfLiteTensor* input_class_predictions =
+      GetInput(context, node, kInputTensorClassPredictions);
+  const TfLiteTensor* input_anchors =
+      GetInput(context, node, kInputTensorAnchors);
+  TF_LITE_ENSURE_EQ(context, NumDimensions(input_box_encodings), 3);
+  TF_LITE_ENSURE_EQ(context, NumDimensions(input_class_predictions), 3);
+  TF_LITE_ENSURE_EQ(context, NumDimensions(input_anchors), 2);
+
+  TF_LITE_ENSURE_EQ(context, NumOutputs(node), 4);
+  const int num_boxes = input_box_encodings->dims->data[1];
+  const int num_classes = op_data->num_classes;
+
+  // Scratch tensors
+  context->RequestScratchBufferInArena(context, num_boxes,
```

Comment:
RequestScratchBufferInArena will request a temporary buffer which will be gone at the end of the given stage. This means requesting scratch buffers in prepare and referencing them from OpData is invalid, since they will have been deallocated before calls to invoke. Please use
buffer = context->AllocatePersistentBuffer(
          context, size_in_bytes). See an example here https://github.com/tensorflow/tensorflow/blob/7ca0373ad3bfeec067c9c604c0973f8455473fdc/tensorflow/lite/micro/kernels/conv.cc#L142

Reply:
I think it is only scratch tensors and temporary buffers (to get rid of std::vector) that are using RequestScratchBufferInArena. And I don't think they need to be preserved at the end of given stage. I have added scratch indexes (*_idx) in OpData as is needed when using RequestScratchBufferInArena. 
I have also added pointers in OpData for these buffers that are set in Eval. The only reason for this is to reduce the number of parameters. Apart from that it is not really needed. It will also look more similar to the TFLite version. However they could be removed. I added some comments to clarify this. Please let me know what you think.

Reply:
Good call on the scope.  I incorrectly thought they needed to be persistent.